File: database/models/MedicalRecord/Allergy.js
import mongoose from 'mongoose';

const allergySchema = new mongoose.Schema({
    userId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    allergen: {
        type: String,
        required: true
    },
    type: {
        type: String,
        enum: ['Thực phẩm', 'Thuốc', 'Môi trường', 'Khác'],
        required: true
    },
    severity: {
        type: String,
        enum: ['Nhẹ', 'Trung bình', 'Nặng', 'Nguy hiểm đến tính mạng'],
        required: true
    },
    symptoms: {
        type: [String],
        required: true
    },
    firstOccurrence: {
        type: Date
    },
    lastOccurrence: {
        type: Date
    },
    treatment: {
        type: String
    },
    notes: {
        type: String
    },
    isActive: {
        type: Boolean,
        default: true
    }
}, {
    timestamps: true
});

const Allergy = mongoose.model('Allergy', allergySchema);

export default Allergy; 

File: mainflutter/lib/models/allergy.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'allergy.freezed.dart';
part 'allergy.g.dart';

@freezed
class Allergy with _$Allergy {
  const factory Allergy({
    @JsonKey(name: '_id') required String id,
    required String allergen,
    required String type,
    required String severity,
    required List<String> symptoms,
    @JsonKey(fromJson: _dateTimeFromJson, toJson: _dateTimeToJson) DateTime? firstOccurrence,
    @JsonKey(fromJson: _dateTimeFromJson, toJson: _dateTimeToJson) DateTime? lastOccurrence,
    String? treatment,
    String? notes,
    @Default(true) bool isActive,
  }) = _Allergy;

  factory Allergy.fromJson(Map<String, dynamic> json) => _$AllergyFromJson(json);
}

DateTime? _dateTimeFromJson(String? date) {
  if (date == null) return null;
  return DateTime.tryParse(date);
}

String? _dateTimeToJson(DateTime? date) {
  if (date == null) return null;
  return date.toIso8601String();
} 